pkg/cli/secretsencrypt/secrets_encrypt.go

```diff
@@ -0,0 +1,224 @@
+package secretsencrypt
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"text/tabwriter"
+
+	"github.com/erikdubbelboer/gspt"
+	"github.com/rancher/k3s/pkg/cli/cmds"
+	"github.com/rancher/k3s/pkg/clientaccess"
+	"github.com/rancher/k3s/pkg/daemons/config"
+	"github.com/rancher/k3s/pkg/server"
+	"github.com/rancher/k3s/pkg/version"
+	"github.com/urfave/cli"
+)
+
+func commandPrep(app *cli.Context, cfg *cmds.Server) (config.Control, *clientaccess.Info, error) {
+	var controlConfig config.Control
+	var err error
+	// hide process arguments from ps output, since they may contain
+	// database credentials or other secrets.
+	gspt.SetProcTitle(os.Args[0] + " encrypt")
+
+	nodeName := app.String("node-name")
+	if nodeName == "" {
+		nodeName, err = os.Hostname()
+		if err != nil {
+			return controlConfig, nil, err
+		}
+	}
+
+	os.Setenv("NODE_NAME", nodeName)
+
+	controlConfig.DataDir, err = server.ResolveDataDir(cfg.DataDir)
+	if err != nil {
+		return controlConfig, nil, err
+	}
+	if cmds.ServerConfig.ServerURL == "" {
+		cmds.ServerConfig.ServerURL = "https://127.0.0.1:6443"
+	}
+
+	if cmds.ServerConfig.Token == "" {
+		fp := filepath.Join(controlConfig.DataDir, "token")
+		tokenByte, err := ioutil.ReadFile(fp)
+		if err != nil {
+			return controlConfig, nil, err
+		}
+		controlConfig.Token = string(bytes.TrimRight(tokenByte, "\n"))
+	} else {
+		controlConfig.Token = cmds.ServerConfig.Token
+	}
+	controlConfig.EncryptForce = cfg.EncryptForce
+	info, err := clientaccess.ParseAndValidateTokenForUser(cmds.ServerConfig.ServerURL, controlConfig.Token, "node")
+	if err != nil {
+		return controlConfig, nil, err
+	}
+	return controlConfig, info, nil
+}
+
+func Enable(app *cli.Context) error {
+	var err error
+	if err = cmds.InitLogging(); err != nil {
+		return err
+	}
+	_, info, err := commandPrep(app, &cmds.ServerConfig)
+	if err != nil {
+		return err
+	}
+	b, err := json.Marshal(server.EncryptionRequest{Enable: true})
+	if err != nil {
+		return err
+	}
+	if err = info.Put("/v1-"+version.Program+"/encrypt/enable", b); err != nil {
+		return err
+	}
+	fmt.Println("secrets-encryption enabled, after server restart run:")
+	fmt.Println("kubectl get secrets --all-namespaces -o json | kubectl replace -f -")
```

Comment:
Wait, are we leaving the actual rotation of encrypted data as an exercise for the reader? I though we were going to manage this bit for them. Having it round-trip through kubectl seems non-optimal.

Reply:
Yes, so we are in the same situation with the enable/disable as the "reencrypt" step. We need to restart the k3s server then, "reread" all the secrets.

Reply:
yeah, I think we should handle rewriting secrets for both the rotate and enable/disable state. Especially since enable/disable are just special cases of secret rotation that involve moving the identity key into or our of primary position. We shouldn't ask the user to effect the rewrite themselves.

Reply:
Reworked error handling. Now the server generates a unique integer and logs that + the error. The CLI simply reports that integer to the user and are promted to go look at the server logs.